### PR TITLE
[Merged by Bors] - feat: technical debts compare to previous week

### PR DESCRIPTION
**Pull request by adomani**
Adds a comparison with the statistics on the version of mathlib from the previous week.

Also renames `tech_debt_counters` to `tech_debt_metrics`.

[Zulip discussion](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/Technical.20Debt.20Counters/near/445055463)

Right now, it reports the following:

|Current number|Change|Type|
|-:|:-:|:-|
|7324|-28|porting notes|
|51|0|backwards compatibility flags|
|90|1|skipAssignedInstances flags|
|141|-1|adaptation notes|
|364|5|disabled simpNF lints|
|168|-12|disabled deprecation lints|
|1936|3|erw|
|961|0|documentation nolint entries|
|30|0|missing module docstrings|
|95|0|large files|
|147|-4|non-test files with autoImplicit true|
|41|0|`Init` files|
|5326|-3|total LoC in `Init` files|

<details><summary>

Changed 'Init' lines by file

</summary>

|LoC|Change|File|
|-:|:-:|-|
|489|0|Mathlib/Init/Algebra/Classes.lean|
|248|-1|Mathlib/Init/Align.lean|
|14|0|Mathlib/Init/Classes/Order.lean|
|37|0|Mathlib/Init/Classical.lean|
|85|0|Mathlib/Init/Control/Combinators.lean|
|264|0|Mathlib/Init/Control/Lawful.lean|
|172|0|Mathlib/Init/Core.lean|
|19|-1|Mathlib/Init/Data/Bool/Basic.lean|
|154|-3|Mathlib/Init/Data/Bool/Lemmas.lean|
|12|0|Mathlib/Init/Data/Buffer/Parser.lean|
|27|0|Mathlib/Init/Data/Fin/Basic.lean|
|122|-1|Mathlib/Init/Data/Int/Basic.lean|
|15|-1|Mathlib/Init/Data/Int/DivMod.lean|
|18|-2|Mathlib/Init/Data/Int/Lemmas.lean|
|100|0|Mathlib/Init/Data/Int/Order.lean|
|70|2|Mathlib/Init/Data/List/Basic.lean|
|65|0|Mathlib/Init/Data/List/Instances.lean|
|182|-1|Mathlib/Init/Data/List/Lemmas.lean|
|44|0|Mathlib/Init/Data/Nat/Basic.lean|
|14|0|Mathlib/Init/Data/Nat/Div.lean|
|132|0|Mathlib/Init/Data/Nat/GCD.lean|
|721|0|Mathlib/Init/Data/Nat/Lemmas.lean|
|11|0|Mathlib/Init/Data/Nat/Notation.lean|
|24|0|Mathlib/Init/Data/Option/Basic.lean|
|16|0|Mathlib/Init/Data/Option/Init/Lemmas.lean|
|67|0|Mathlib/Init/Data/Option/Lemmas.lean|
|45|0|Mathlib/Init/Data/Ordering/Basic.lean|
|61|0|Mathlib/Init/Data/Ordering/Lemmas.lean|
|12|0|Mathlib/Init/Data/Prod.lean|
|65|0|Mathlib/Init/Data/Quot.lean|
|17|0|Mathlib/Init/Data/Rat/Basic.lean|
|22|0|Mathlib/Init/Data/Sigma/Basic.lean|
|74|0|Mathlib/Init/Data/Sigma/Lex.lean|
|252|0|Mathlib/Init/Function.lean|
|673|3|Mathlib/Init/Logic.lean|
|27|0|Mathlib/Init/Meta/WellFoundedTactics.lean|
|469|0|Mathlib/Init/Order/Defs.lean|
|173|0|Mathlib/Init/Order/LinearOrder.lean|
|43|0|Mathlib/Init/Quot.lean|
|222|0|Mathlib/Init/Set.lean|
|49|2|Mathlib/Init/ZeroOne.lean|
|5326|-3|total|
</details>

Current commit [a59524cd3b](https://github.com/leanprover-community/mathlib4/commit/a59524cd3bf72a1f393255869dd580a4c9d7db2d)
Reference commit [c95c52ea4c](https://github.com/leanprover-community/mathlib4/commit/c95c52ea4cf299397d7771bacfcfcadaf084052b)

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Review by grunweg — APPROVED** [review]
I don't speak awk; the remaining changes look good to me. (I have some minor tweaks to suggest, all of them optional.) Thanks for doing this!

I'll just approve, to let somebody else take a look at the awk logic.

File: scripts/technical-debt-counters.sh [hunk 1]
```diff
@@ -1,7 +1,35 @@
 #!/usr/bin/env bash
 
-# This script quantifies some of the technical debt in mathlib4.
+# `./scripts/et.sh` returns a tally of some technical debts in current Mathlib,
```
[the remark this answers is not in the document]
Comment (adomani) [reply scripts/technical-debt-counters.sh]: Yes, this was a "development name", good catch!  I also like your suggestion and renamed the file with the script.

File: scripts/technical-debt-counters.sh [hunk 2]
```diff
@@ -1,7 +1,35 @@
...
+
+if [ -n "${1}" ]; then
+  currCommit="${1}"
+else
+  currCommit="$(git rev-parse HEAD)"
+fi
+
+shift
+
+if [ -n "${1}" ]; then
```
[the remark this answers is not in the document]
Comment (adomani) [reply scripts/technical-debt-counters.sh]: Sure, I removed `shift`.

File: scripts/technical-debt-counters.sh [hunk 3]
```diff
@@ -1,7 +1,35 @@
...
+  refCommit="${1}"
+else
+  refCommit="$(git log --pretty=%H --since="$(date -I -d 'last week')" | tail -1)"
```
[the remark this answers is not in the document]
Comment (adomani) [reply scripts/technical-debt-counters.sh]: Sure, I can use `tail -n -1` (I am not really sure about the difference anyway).

File: scripts/technical-debt-counters.sh [hunk 4]
```diff
@@ -22,21 +50,46 @@ for i in ${!titlesAndRegexes[@]}; do
...
-initFiles="$(git ls-files '**/Init/*.lean' | xargs wc -l)"
+initFiles="$(git ls-files '**/Init/*.lean' | xargs wc -l | sed 's=^ *==')"
 
-printf '|%s | %s |\n\n' "$(printf '%s' "${initFiles}" | wc -l)" "\`Init\` files"
+printf '%s|%s\n' "$(printf '%s' "${initFiles}" | wc -l)" "\`Init\` files"
+printf '%s|%s\n\n' "$(printf '%s\n' "${initFiles}" | grep total | sed 's= total==')"  'total LoC in `Init` files'
 
-printf '```spoiler %s %s\n%s\n```\n' "$(printf '%s\n' "${initFiles}" | grep total)"  "LoC in 'Init' files" "$(
-    printf '%s\n' "${initFiles}" | awk 'BEGIN{print"|LoC|File|\n|-:|-|"} {printf("|%s|%s|\n", $1, $2)}'
+printf $'```spoiler \'Init\' file by file\n%s\n```\n' "$(
```
[the remark this answers is not in the document]
Comment (adomani) [reply scripts/technical-debt-counters.sh]: Good suggestion: I used "Changed 'Init' lines by file", hoping that it is even clearer!

**Review by adomani — COMMENTED** [review]
Michael, as usual, thanks a lot!  I implemented all your suggestions.  I also added some comment to the `awk` script.

File: scripts/technical-debt-metrics.sh [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+  awk -F'|' '
+    # if the first "word" is a number, then we write the 1st entry and compare it with the 4th
+    ($1+0 == $1) { printf("|%s|%s|%s|\n", $1, $1-$4, $2) }
+    # otherwise, the line is a "formatting" line, so we simply print it unchanged until we find `@`
+    !($1+0 == $1) {
+      found=0
+      for(i=1; i<=NF; i++) {
+        if ($i == "@") { found=1 }
+        if (found == "0") { printf("%s|", $i) }
+      }
```
Comment (bryangingechen) [inline scripts/technical-debt-metrics.sh:93]: I don't really know awk either, but could we use a `break` statement here instead?
```suggestion
      for(i=1; i<=NF; i++) {
        if ($i == "@") break
        printf("%s|", $i)
      }
```

**Review by bryangingechen — APPROVED** [review]
Otherwise LGTM.
bors d+

File: scripts/technical-debt-metrics.sh (hunk 5 above)
Comment (adomani) [reply scripts/technical-debt-metrics.sh]: Bryan thanks, this works very well!

**Pull request closed without merging**